test/index.test.ts

```diff
@@ -14,6 +36,81 @@ interface MockUser {
   email: string;
 }
 
+type ReturnTypeFromStage<
+  T extends (context: any) => any,
+  TStage extends AllStageKeys
+> = TStage extends AllAsyncStageKeys
+  ? PromiseResolveOrSync<ReturnType<T>>
+  : ReturnType<T>;
+
+async function HTPipeTest<
+  TPipe extends OptionallyHasInitPreContext<any, any> &
+    OptionallyHasSanitizeParams<any, any> &
+    OptionallyHasSanitizeBody<any, any> &
+    MightHavePreAuthorize<any, any> &
+    OptionallyHasAttachData<any, any> &
+    MightHaveFinalAuthorize<any, any> &
+    OptionallyHasDoWork<any, any> &
+    MightHaveRespond<any, any> &
+    MightHaveSanitizeResponse<any, any>,
+  TPipeIn extends TPipeInExpected,
+  TPipeOut extends TPipeOutExpected,
+  TStage extends AllStageKeys,
+  TLifecycleStage extends TPipe extends HasInitPreContext<any, any>
+    ? TPipe['initPreContext']
+    : TPipe extends HasSanitizeParams<any, any>
+    ? TPipe['sanitizeParams']
+    : TPipe extends HasSanitizeBody<any, any>
+    ? TPipe['sanitizeBody']
+    : TPipe extends HasPreAuthorize<any, any>
+    ? TPipe['preAuthorize']
+    : TPipe extends HasAttachData<any, any>
+    ? TPipe['attachData']
+    : TPipe extends HasFinalAuthorize<any, any>
+    ? TPipe['finalAuthorize']
+    : TPipe extends HasDoWork<any, any>
+    ? TPipe['doWork']
+    : TPipe extends HasRespond<any, any>
+    ? TPipe['respond']
+    : TPipe extends HasSanitizeResponse<any, any>
+    ? TPipe['sanitizeResponse']
+    : never,
+  TValid extends TPipeInExpected extends Parameters<TLifecycleStage>[0]
+    ? Parameters<TLifecycleStage>[0] extends TPipeInExpected
+      ? TPipeOutExpected extends ReturnTypeFromStage<TLifecycleStage, TStage>
+        ? ReturnTypeFromStage<TLifecycleStage, TStage> extends TPipeOutExpected
+          ? true
+          : never
+        : never
+      : never
+    : never,
+  TPipeInExpected = TPipeIn,
+  TPipeOutExpected = TPipeOut
+>(
+  pipe: TPipe,
+  lifecycleStage: TStage,
+  pipeIn: TPipeIn,
+  pipeOut: TPipeOut,
+  valid: TValid,
```

Comment:
I don't think we need pipeInExpected and pipeOutExpected FUNCTION parameters ... we just needed the optional TYPE parameters earlier.  99% of the time, it won't even be used, and it'll just be TPipeIn and TPipeOut respectively
- [ ] remove function params.

Reply:
also, on the javascript side, can't you make valid into `valid: TValid = true` so that you don't have to pass it in each time?
- [ ] valid default to true

Reply:
done
valid can't be true by default i tried that
the problem is it gives error when i try to set valid = true
error is:
Type 'true' is not assignable to type 'TValid'.
  'true' is assignable to the constraint of type 'TValid', but 'TValid' could be instantiated with a different subtype of constraint 'true'.

Reply:
bastard typescript